Templates Endpoint: Add `resolved` query arg to return only relevant templates

ockham opened the pull request:
## Description
Implements https://github.com/WordPress/gutenberg/pull/21877/files#r416075231 and unblocks #21877.

This PR changes `gutenberg_find_template_post_and_parts()` to accept a `$template_type` arg, and make the `$template_hierarchy` arg optional. `$template_hierarchy` will be used as a hint if it's present (e.g. when called from a `{$type}_template` hook); if it isn't, the template hierarchy will be looked up for `$template_type`.

This allows writing the code in `edit-site-page.php` in a more succinct way. More importantly, it allows us to add a `rest_wp_template_query` filter for the `wp_template` collection endpoint to accept a `resolved` bool arg that will cause the endpoint to reflect the behavior of the template resolution algorithm that's used to populate the `$settings[ 'templateIds' ]` ([see](https://github.com/WordPress/gutenberg/pull/21877/files#r415995080)):

> Published templates are always used first and mean the template was customized. Auto drafts are only used when the templates haven't been customized. We should preferably use the latest ones

## How has this been tested?

As always, make sure that the Full-Site Editing experiment and demo templates are enabled.

We'll compare the results the `wp_template` collections REST API endpoint gives to `$settings['templateIds']`. To that end, apply the following patch and re-build the app (`npm run build`):

```diff
diff --git a/packages/edit-site/src/index.js b/packages/edit-site/src/index.js
index e6dec20771..ea72bee4c9 100644
--- a/packages/edit-site/src/index.js
+++ b/packages/edit-site/src/index.js
@@ -26,6 +26,7 @@ export function initialize( id, settings ) {
        if ( process.env.GUTENBERG_PHASE === 2 ) {
                __experimentalRegisterExperimentalCoreBlocks( settings );
        }
+       console.log( 'settings', settings );
        render( <Editor settings={ settings } />, document.getElementById( id ) );
 }

```

Next, in wp-admin, navigate to the Appearance > Templates section, and create a few dummy templates whose names are **not** actually part of the template hierarchy (e.g. `some-template` etc).

Now go to Full-Site Editing and open your browser console. You should see the output of the `settings` variable there. Compare the `templateIds` property to the output of the following commands you run in the console (twice each!). Icons :heavy_check_mark: and :x: indicate whether the output is expected to match `settings.templateIds`, or not.

- :x: `wp.data.select('core').getEntityRecords('postType', 'wp_template' )` should return the list of all `wp_template` CPTs, including those that are not part of the template hierarchy.
- :heavy_check_mark: `wp.data.select('core').getEntityRecords('postType', 'wp_template', { resolved: true } )` should return the list of templates that match the template hierarchy.
- :x: `wp.data.select('core').getEntityRecords('postType', 'wp_template', { slug: 'some-template' } )` should return the `wp_template` part whose slug is `some-template`.
- :heavy_check_mark: `wp.data.select('core').getEntityRecords('postType', 'wp_template', { slug: 'some-template', resolved: true } )` should return an empty array, as `some-template` is not part of the template hierarchy.

## Types of changes
Add an arg to an existing endpoint; plus a minor refactor.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

ockham committed:
Templates Endpoint: Add resolved filter

ockham committed:
Template Loader: Search by template type, make template hierarchy optional

ockham committed:
Add TODO comments about future optimization

epiqueras commented on lib/template-loader.php:
```diff
@@ -191,19 +191,30 @@ function create_auto_draft_for_template_part_block( $block ) {
- * @param string[] $template_hierarchy The current template hierarchy, ordered by priority.
+ * Accepts an optional $template_hierarchy argument as a hint.
+ *
+ * @param string   $template_type The current template type.
+ * @param string[] $template_hierarchy (optional) The current template hierarchy, ordered by priority.
```
```suggestion
 * @param string   $template_type      The current template type.
 * @param string[] $template_hierarchy (optional) The current template hierarchy, ordered by priority.
```

epiqueras commented on lib/template-loader.php:
```diff
@@ -223,6 +234,10 @@ function gutenberg_find_template_post_and_parts( $template_hierarchy ) {
 
 	$current_template_post = $template_query->have_posts() ? $template_query->next_post() : null;
 
+	// TODO: We could consider moving the following logic to a hook like `posts_results` that is run
+	// after query results have been fetched. That might allow us to absorb filtering logic
+	// (as e.g. found in `filter_rest_wp_template_query`) into the generic querying logic.
```
This comment is confusing. How would moving this into a hook allow for `filter_rest_wp_template_query` to be in the "generic querying logic."?

ockham replied:
What I meant is to move the auto-draft creation logic into a `post_results` filter, and to filter the results returned by the query to only return the template(s) that are relevant for the requested template hierarchy -- similar to what `filter_rest_wp_template_query` does in this PR.

epiqueras replied:
The auto-drafts are not created for the templates that are returned by the query. They are created for spots in the template hierarchy that don't have a returned template but have a theme file. You would have to create the auto-drafts before making the query so that they show up, but give preference to non-auto-drafts. That could work, but I doubt it will be more straightforward.

ockham replied:
Yeah, I think we're basically on the same page. <y thinking was basically:

- We tap into the `post_results` hook to watch out for a query for one (or many) `wp_template`(s), possibly with a `post_name` or `post_name__in` param, akin to this: https://github.com/WordPress/gutenberg/blob/8676302f5cc27838d0c7eb586d2384041ece0ed2/lib/template-loader.php#L212-L222
- In our filter, we create the auto-drafts from theme files, and if there's one that matches the template hierarchy better than our query result, we return that -- that's basically https://github.com/WordPress/gutenberg/blob/8676302f5cc27838d0c7eb586d2384041ece0ed2/lib/template-loader.php#L224-L280

While it mangles the query logic quite a bit, it's arguable that we'd pretty much always like to perform these steps whenever a template is queried (whether for rendering or through the REST API). I think it might be worth experimenting with.

epiqueras replied:
We could also just call the new functions directly for rendering.

epiqueras commented on lib/templates.php:
```diff
@@ -155,3 +155,55 @@ function gutenberg_render_template_list_table_column( $column_name, $post_id ) {
+function filter_rest_wp_template_query( $args, $request ) {
+	if ( $request['resolved'] ) {
+		$template_ids  = array( 0 ); // Return nothing by default (the 0 is needed for `post__in`).
+		$template_type = $request['slug'] ? array( $request['slug'] ) : get_template_types();
```
This gets shadowed before it's used.

ockham replied:
Uh, half-reverted refactor. Good spot, I'll fix.

ockham replied:
7d72ff7

epiqueras commented on lib/templates.php:
```diff
@@ -155,3 +155,55 @@ function gutenberg_render_template_list_table_column( $column_name, $post_id ) {
+				continue;
+			}
+
+			// TODO: gutenberg_find_template_post_and_parts() internally performs a query.
+			// This means that there's room for optimization, and we might be able to
+			// generalize the logic found here into something that runs upon every `wp_template`
+			// query (not just those coming from the REST API).
```
Why would we want that? We still need a way to use the regular querying logic.

ockham replied:
I was thinking of adding some kind of flag, similar to the `resolved` bool we're adding to the REST API endpoint in this PR. The same technique might not carry over 1:1 to `WP_Query`, but maybe we could add e.g. a dummy meta field.

I think that we _could_ consider manipulating _all_ queries for `wp_template`. If you look at the querying and auto-draft creation logic in `gutenberg_find_template_post_and_parts`, and where it's used, it's arguable that we'll always want to create those auto-drafts when querying a `wp_template`.

epiqueras replied:
But, you need to query the templates to figure out which ones you need to create auto-drafts for. The only way I see to get around that is to generate auto-drafts for all files before queries are made. Still, then we have to make sure we always update them when the underlying files change and that we give priority to actual customized non-auto-draft counterparts. All in all, I don't see it being much more straightforward, and we introduce a lot of magic into the very standardized functionality of `WP_Query`.

ockham replied:
> But, you need to query the templates to figure out which ones you need to create auto-drafts for. The only way I see to get around that is to generate auto-drafts for all files before queries are made. 

That depends on the query, doesn't it? If there's no `post_name` or `post_name__in` param, then that's true -- but that means it'd be pretty similar to what we have now in `edit-site-page.php` or in the new REST API filter I'm adding here -- where we just loop over template hierarchies.

> Still, then we have to make sure we always update them when the underlying files change and that we give priority to actual customized non-auto-draft counterparts.

That should be feasible by splitting what we now have in `gutenberg_find_template_post_and_parts` along sensible lines (like I sketched [here](https://github.com/WordPress/gutenberg/pull/21981#discussion_r423962066)), no?

> All in all, I don't see it being much more straightforward, and we introduce a lot of magic into the very standardized functionality of `WP_Query`.

Yeah, I think the magic vs. expected behavior is the biggest concern.

epiqueras replied:
Yeah, I don't see it as a definite net improvement.

We can explore it, but I wouldn't prioritize it.

ockham committed:
PHPDoc whitespace

Co-authored-by: Enrique Piqueras <epiqueras@users.noreply.github.com>

ockham committed:
Fix filter_rest_wp_template_query

ockham commented:
Addressed feedback. If the TODO comments are the main concern, I'm happy to revert them. Also happy to demo what I meant with them in a separate PR (since I'd rather not block this one on it(.

epiqueras reviewed (approved):
Let's remove the TODOs, and merge.

ockham committed:
Revert "Add TODO comments about future optimization"

This reverts commit 6ca25eac8c060c69e32ff946fc586a2570aae749.

ockham committed:
phpcbf
